### PR TITLE
Generate AppImage

### LINKED ISSUE #2024: Provide official AppImage builds for Linux
Please provide PowerShell in the AppImage format for Linux.

[AppImage](http://appimage.org/) is a packaging format for Linux that lets users download one single file, set the executable bit and run the application (here: PowerShell). Unlike other solutions, no special runtime infrastructure must be installed on the target system(s).

PowerShell

An unofficial PowerShell AppImage has been available on https://bintray.com/probono/AppImages/PowerShell#files. It has been tested on
- CentOS-7.0-1406-x86_64-GnomeLive.iso
- CentOS-7-x86_64-LiveGNOME-1511.iso
- Chromixium-1.5-amd64.iso
- Fedora-Live-Workstation-x86_64-22-3.iso
- Fedora-Live-Workstation-x86_64-23-10.iso
- SL-72-x86_64-2016-02-03-LiveDVDgnome.iso
- debian-live-8.0.0-amd64-xfce-desktop+nonfree.iso
- debian-live-8.4.0-amd64-gnome-desktop.iso
- elementary_OS_0.3_freya_amd64.iso
- kali-linux-2.0-amd64.iso
- kali-linux-2016.1-amd64.iso
- kubuntu-14.04.4-desktop-amd64.iso
- kubuntu-15.04-desktop-amd64.iso
- kubuntu-16.04-desktop-amd64.iso
- linuxmint-17.3-cinnamon-64bit.iso
- neon-devedition-gitunstable-20160814-0806-amd64.iso
- netrunner-17-64bit.iso
- ubuntu-14.04.1-desktop-amd64.iso
- ubuntu-16.04-desktop-amd64.iso
- ubuntu-gnome-16.04-desktop-amd64.iso
- ubuntu-mate-16.04-desktop-amd64.iso
- xubuntu-16.04-desktop-amd64.iso

To try it, run

```
PS="PowerShell-6.0.0.alpha.9-x86_64.AppImage"
wget -c "https://bintray.com/probono/AppImages/download_file?file_path=$PS" -O "$PS"
chmod a+x "$PS"
./"$PS"
```

@TravisEz13 mentioned that the AppImage works for him and @andschwa mentioned that he likes the idea of having one format for all Linux distributions.

The script that generated the AppImage is [here](https://github.com/probonopd/AppImages/blob/master/recipes/powershell/Recipe). Feel free to use it, e.g. as part of the automated Travis CI build process. I am happy to help.

### PR DESCRIPTION
**Pull request by probonopd**
Closes #2024

[AppImage](http://appimage.org/) is a packaging format for Linux that lets users download one single file, set the executable bit and run the application (here: PowerShell). Unlike other solutions, no special runtime infrastructure must be installed on the target system(s).

This pull request, when merged, will package PowerShell as an AppImage and upload it.

### CODE REVIEW

**Comment by probonopd** [comment]
Here is the log of a build in which the AppImage actually gets generated (it does not in PRs, as do the other packages): https://travis-ci.org/probonopd/PowerShell/builds/154322840

File: docs/installation/linux.md [hunk 1]
```diff
@@ -2,14 +2,15 @@ Package installation instructions
 =================================
 
 Supports [Ubuntu 14.04][u14], [Ubuntu 16.04][u16],
-[CentOS 7][cos], and [OS X 10.11][osx].
+[CentOS 7][cos], [most Linux distributions (AppImage)][lai], and [OS X 10.11][osx].
```
Comment (TravisEz13) [inline docs/installation/linux.md]: Can we say many instead of most?  And add a link to AppImage for more info.

File: README.md [hunk 2]
```diff
@@ -27,21 +27,24 @@ You can download and install a PowerShell package for any of the following platf
 | Ubuntu 16.04                   | [.deb][rl-ubuntu16]    | [Instructions][in-ubuntu16]   |
 | Ubuntu 14.04                   | [.deb][rl-ubuntu14]    | [Instructions][in-ubuntu14]   |
 | CentOS 7                       | [.rpm][rl-centos]      | [Instructions][in-centos]     |
+| Most Linux distributions       | [.AppImage][rl-ai]     | [Instructions][in-appimage]   |
 | OS X 10.11                     | [.pkg][rl-osx]         | [Instructions][in-osx]        |
-| Docker                         |                        | [Instructions] [in-docker]    |
+| Docker                         |                        | [Instructions][in-docker]     |
```
Comment (TravisEz13) [inline README.md]: Thanks for the fix

File: README.md [hunk 3]
```diff
@@ -27,21 +27,24 @@ You can download and install a PowerShell package for any of the following platf
 | Ubuntu 16.04                   | [.deb][rl-ubuntu16]    | [Instructions][in-ubuntu16]   |
 | Ubuntu 14.04                   | [.deb][rl-ubuntu14]    | [Instructions][in-ubuntu14]   |
 | CentOS 7                       | [.rpm][rl-centos]      | [Instructions][in-centos]     |
+| Most Linux distributions       | [.AppImage][rl-ai]     | [Instructions][in-appimage]   |
```
Comment (TravisEz13) [inline README.md]: Many comment again.

File: tools/appimage.sh [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+# xubuntu-16.04-desktop-amd64.iso
+
+APP=PowerShell
+LOWERAPP=${APP,,}
+
+mkdir -p ./$APP/$APP.AppDir/usr/lib
+
+cd ./$APP/
+
+wget -q https://github.com/probonopd/AppImages/raw/master/functions.sh -O ./functions.sh
```
Comment (TravisEz13) [inline tools/appimage.sh]: Please add a link the the license, mention the license type.

File: tools/appimage.sh [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+Exec=powershell
+Keywords=shell;prompt;command;commandline;cmd;
+Icon=powershell
+Type=Application
+Categories=System;TerminalEmulator;
+StartupNotify=true
+Terminal=true
+EOF
+
+wget -c "https://github.com/PowerShell/PowerShell/raw/master/assets/Powershell_256.png" -O $LOWERAPP.png
```
Comment (TravisEz13) [inline tools/appimage.sh]: This will be local, you don't need to get it.

**Comment by TravisEz13** [comment]
Can you do an interactive rebase so that there is one change per file (or less?)

File: docs/installation/linux.md (hunk 1 above)
Comment (probonopd) [inline docs/installation/linux.md]: Yes, done

File: README.md (hunk 3 above)
Comment (probonopd) [inline README.md]: Done

File: tools/appimage.sh (hunk 4 above)
Comment (probonopd) [inline tools/appimage.sh]: Done

File: tools/appimage.sh (hunk 5 above)
Comment (probonopd) [inline tools/appimage.sh]: Done

**Comment by probonopd** [comment]
Thanks for the review @TravisEz13.

> Can you do an interactive rebase so that there is one change per file (or less?)

As in
```
git rebase -i ${HASH}
git config --global push.default simple
git push -f
```
Correct?

Wouldn't it be less dangerous if you would use the GitHub [pull request merge squashing](https://help.github.com/articles/about-pull-request-merge-squashing/) setting?

**Comment by TravisEz13** [comment]
@probonopd , I've assigned it to @andschwa  for final review.  He can decide if squashing is acceptable.

**Comment by TravisEz13** [comment]
AppVeyor failed to build FullCLR, restarting to see if it was an intermittent issue as I could not find the root cause:  https://ci.appveyor.com/project/PowerShell/powershell/build/6.0.0-alpha.9-183

**Comment by probonopd** [comment]
Now it says "All checks have passed". Will file an issue regarding the failed upload.

File: tools/appimage.sh [hunk 6]
```diff
@@ -0,0 +1,146 @@
+#!/bin/bash
+
+# The MIT License (MIT)
+# Copyright (c) 2016 Simon Peter
+
```
Comment (TravisEz13) [inline tools/appimage.sh]: Software being committed to this repos should me copyright Microsoft Corporation, per the CLA, and me already have a CLA.

File: tools/appimage.sh (hunk 4 above)
Comment (TravisEz13) [inline tools/appimage.sh]: I was saying here you should mention that functions.sh is covered by the MIT license, and the copyright it is covered under.

---
In reply to: [75963644](https://github.com/PowerShell/PowerShell/pull/2027#discussion_r75963644) [](ancestors = 75963644)

File: tools/appimage.sh (hunk 6 above)
Comment (TravisEz13) [inline tools/appimage.sh]: If this was based on existing script covered by an existing MIT license, Please attribute here as well, give the location of the license (and mention the type).

---
In reply to: [75977883](https://github.com/PowerShell/PowerShell/pull/2027#discussion_r75977883) [](ancestors = 75977883)

Comment (probonopd) [inline tools/appimage.sh]: Yes, this is based on this [existing script](https://github.com/probonopd/AppImages/blob/master/LICENSE) which comes from the [probonopd/AppImages](https://github.com/probonopd/AppImages/) project which in its entirety [is MIT licensed](https://github.com/probonopd/AppImages/blob/master/LICENSE). Hence I don't think that "copyright Microsoft Corporation" is applicable here, please advise.

Comment (TravisEz13) [inline tools/appimage.sh]: Add something to the effect of the following replacing the placeholders with the actual values.  Please be sure to mention that the script itself and functions.sh.

```
This code is based on <description/name of project>,  <insert copyright notice>
<location of original code>
This license of this code is an MIT License, see <url to license>
```

**Comment by andschwa** [comment]
This looks good to me. Having not used AppImage, I can't comment as to the actual setup, but it clearly works 😄 

@TravisEz13 are you okay with licensing/copyright?

@probonopd do you know what happened at the [end of the log](https://travis-ci.org/probonopd/PowerShell/builds/154322840#L1366) of the build you posted?

**Comment by andschwa** [comment]
I'm running a non-PR [build here](https://travis-ci.org/andschwa/PowerShell/builds/155402372) to test.

**Comment by andschwa** [comment]
Yeah, can't quite merge. The build failed to run the Pester tests afterwards. @probonopd is it a `cd` issue?

**Comment by probonopd** [comment]
If you are referring to [this](https://travis-ci.org/probonopd/PowerShell/builds/154322840#L1366-L1375), what is it supposed to do?

And yes, could be necessary to `cd ..`, will try that.

**Comment by andschwa** [comment]
It doesn't appear to be a problem related to the AppImage. I tested both saving and restoring the location; and not doing the AppImage generation until after the Pester tests. I'm testing a rebase now.

**Comment by probonopd** [comment]
Continued in #2082, hence closing here.

**Pull request closed without merging**